generative_models/cGAN_PyTorch/main.py

```diff
@@ -0,0 +1,172 @@
+    fix_labels = Variable(torch.LongTensor(fix_labels)).to('cuda:0')
+
+    G_losses, D_losses = train(args,generator,discriminator,dataloader,loss, img_type, label_type,gen_optimizer,d_optimizer,fix_labels,fix_z)
+
+    # Plotting the loss graph    
+    plt.plot(G_losses, label='Generator')
+    plt.plot(D_losses, label='Discriminator')
+    plt.legend()
+    plt.savefig(args.outdir+"plot.png")
+    plt.show()
+
+    # Making the GIF
+    image = []
+    for i in range(1,args.epochs+1):
+      image.append(imageio.imread(args.outdir+'Image/'+str(i)+'.png'))
+    imageio.mimsave(args.ndata+'.gif', image, fps=5)
+
+
+def sample_image(args, z, labels, batches_done, generator):
+    """Saves a grid of generated digits ranging from 0 to n_classes"""
+    gen_imgs = generator(z, labels)
+    save_image(gen_imgs.data, args.outdir + "Image/%d.png" % batches_done, nrow=args.num_class, normalize=True)
+
+
+def train(args, generator, discriminator, dataloader, loss, img_type, label_type, gen_optimizer, d_optimizer, fix_label, fix_noise):
+
+    generator.train()
+    discriminator.train()
+    G_losses = []
+    D_losses = []
+
+    for epoch in range(1, args.epochs + 1):
+        G_loss = 0.
+        D_loss = 0.
+        start_time = time.time()
+        for i, data in enumerate(dataloader):
+            (imgs, labels) = data
+
+            batch_size = imgs.shape[0]
+            # print(batch_size)
+            imgs = Variable(imgs.type(img_type)).to(args.device)
+            labels = Variable(labels.type(label_type)).to(args.device)
+
+            # Creating real and fake label for calculation of loss
+            r_label = Variable(img_type(batch_size, 1).fill_(0.9)).to(args.device)
+            f_label = Variable(img_type(batch_size, 1).fill_(0.0)).to(args.device)
+
+            # Training Generator
+
+            gen_optimizer.zero_grad()
+
+            noise = Variable(img_type(np.random.normal(0, 1, (batch_size, args.latent_dim)))).to(args.device)
+            rand_label = Variable(label_type(np.random.randint(0, args.num_class, batch_size))).to(args.device)
+            dis = discriminator(generator(noise, rand_label), rand_label)
+            # print(type(dis),'  ',type(r_label))
+            g_loss = loss(dis, r_label)
+            g_loss.backward()
+            gen_optimizer.step()
+
+            # Training Discriminator
+
+            d_optimizer.zero_grad()
+
+            noise = Variable(img_type(np.random.normal(0, 1, (batch_size, args.latent_dim)))).to(args.device)
+            rand_label = Variable(label_type(np.random.randint(0, args.num_class, batch_size))).to(args.device)
+
+            d_real = discriminator(imgs, labels)
+            loss_real = loss(d_real, r_label)
+
+            d_fake = discriminator(generator(noise, rand_label).detach(), rand_label)
+            # print(d_fake.shape," ",f_label.shape)
+            loss_fake = loss(d_fake, f_label)
+
+            d_loss = 0.5 * (loss_fake + loss_real)
```

Comment:
```suggestion
            d_loss = loss_fake + loss_real
```

Reply:
Do I need to train all these models again and generate the graph and all

Reply:
uploaded new graphs for this commit